diff --git a/src/Nix/TH.hs b/src/Nix/TH.hs
index 87cac23..f77dd47 100644
--- a/src/Nix/TH.hs
+++ b/src/Nix/TH.hs
@@ -15,9 +15,11 @@ import           Data.Set                       ( Set
                                                 )
 import qualified Data.Set                      as Set
 import qualified Data.Text                     as Text
+import           Data.Typeable                  ( cast )
 import           Data.List.NonEmpty             ( NonEmpty(..) )
 import           Data.Maybe                     ( mapMaybe )
 import           Language.Haskell.TH
+import           Language.Haskell.TH.Syntax     ( liftString )
 import           Language.Haskell.TH.Quote
 import           Nix.Atoms
 import           Nix.Expr
@@ -27,8 +29,13 @@ quoteExprExp :: String -> ExpQ
 quoteExprExp s = do
   expr <- case parseNixText (Text.pack s) of
     Failure err -> fail $ show err
-    Success e   -> return e
-  dataToExpQ (const Nothing `extQ` metaExp (freeVars expr)) expr
+    Success e   -> pure e
+  dataToExpQ
+    (const Nothing `extQ` metaExp (freeVars expr) `extQ` (Just . liftText))
+    expr
+ where
+  liftText :: Text.Text -> Q Exp
+  liftText txt = AppE (VarE 'Text.pack) <$> liftString (Text.unpack txt)
 
 quoteExprPat :: String -> PatQ
 quoteExprPat s = do
